File: CivitasRuby/lib/sorpresa_por_jugador.rb
# encoding:utf-8
require_relative 'sorpresa'
module Civitas
  class SorpresaPorJugador < Sorpresa
    def initialize(texto,valor)
      super(texto)
      @valor = valor
    end
    
    def informe(actual, todos)
      Diario.instance.ocurre_evento("Aplicando sorpresa " + self.to_s + 
        " al jugador " + todos[actual].nombre + " de tipo PorJugador")
    end
    
    def aplicar_a_jugador(actual, todos)
      if(super.jugador_correcto(actual,todos))
        informe(actual,todos)
        valor_actual = @valor * (todos.size-1)
        valor_otros = @valor * -1
        
        for i in todos
          if i==todos[actual]
            i.modificar_saldo(valor_actual)
          else
            i.modificar_saldo(valor_otros)
          end
        end       
      end
    end
    
    def to_s
      "Sorpresa: { \n Texto: #{@texto}  \n Tipo: PagarCobrar   \n Valor: #{@valor} }"
    end
  end
end

File: CivitasJava/src/Civitas/SorpresaPorJugador.java
package Civitas;

import java.util.ArrayList;

public class SorpresaPorJugador extends Sorpresa{
    private int valor;
    
    SorpresaPorJugador(int valor, String texto){
        super(texto);
        this.valor=valor;
    }
    
    @Override
    void aplicarAJugador(int actual, ArrayList<Jugador> todos){
       if(jugadorCorrecto(actual, todos)){
            informe(actual, todos);
            int valorActual = this.valor;
            int valorOtros = this.valor;
            String text = " Pagar al jugador " + todos.get(actual).getNombre();
            valorActual = valorActual * (todos.size()-1);
            valorOtros = valorOtros * -1;
            Sorpresa sorpActual = new SorpresaPagarCobrar(valorActual, text);
            Sorpresa sorpOtros =  new SorpresaPagarCobrar(valorOtros, text);

            for(int i=0; i < todos.size(); i++){
                if(i==actual){
                    sorpActual.aplicarAJugador(actual,todos);
                }else{
                    sorpOtros.aplicarAJugador(i, todos);
                }
            }        
        }     
    }
    
    @Override
    public String toString(){
        return "Sorpresa: { \n Texto = " + texto + "\n Tipo = PorJugador" + 
                "\n Valor = " + valor + "}";
    }
}
